libraries/AP_HAL/UARTDriver.cpp

```diff
@@ -226,3 +232,124 @@ void AP_HAL::UARTDriver::log_stats(const uint8_t inst, StatsTracker &stats, cons
 }
 #endif // HAL_LOGGING_ENABLED
 #endif // HAL_UART_STATS_ENABLED
+
+
+/*
+ * Support for logging all incoming/outgoing data from a serial port
+ */
+
+#if HAL_UART_DEBUG_LOGGING_ENABLED
+
+/*
+  log some data for debugging
+*/
+void AP_HAL::UARTDriver::log_data(const uint8_t *data, uint16_t length, bool is_written_data)
+{
+    // first see if we actually do want to log:
+    if ((_last_options & OPTION_LOGGING) == 0) {
+        return;
+    }
+
+    // find logging instance corresponding to this UART:
+    if (loginfo == nullptr) {
+        loginfo = NEW_NOTHROW LogInfo();
+        if (loginfo == nullptr) {
+            return;
+        }
+
+        WITH_SEMAPHORE(loginfo_creation_sem);
+
+        loginfo->instance = instance;
+        loginfo->next = backend_loginfos;
+        backend_loginfos = loginfo;
+
+        if (!log_thread_created) {
+            log_thread_created = true;
+            hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_HAL::UARTDriver::logging_start, void), "uart_log", 4096, AP_HAL::Scheduler::PRIORITY_IO, 0);
+        }
+    }
+
+    // write packet into buffer:
+    struct PACKED Header {
+        enum class Flag {
+            IS_WRITTEN_DATA = 1,
+        };
+        uint32_t magic = 0xEAEF0D0F;
+        uint32_t time_ms;
+        uint16_t length;
+        uint8_t flags;
+    };
+
+    // note that we are assuming we can do this writing atomically.
+    // If there are multiple writers to the serial port that won't be
+    // the case.  We could reserve, write to and commit an iovec
+    // instead of this:
+
+    const auto space_required = sizeof(Header) + length;
+
+    if (loginfo->buf.space() < space_required) {
+        return;
+    }
+
+    Header header;
+    header.time_ms = AP_HAL::millis();
+    header.length = length;
+    if (is_written_data) {
+        header.flags |= uint8_t(Header::Flag::IS_WRITTEN_DATA);
+    }
+
+    loginfo->buf.write((uint8_t*)&header, sizeof(header));
+    loginfo->buf.write(data, length);
+}
+
+// a linked list of backends which are currently logging:
+AP_HAL::UARTDriver::LogInfo *AP_HAL::UARTDriver::backend_loginfos;
+bool AP_HAL::UARTDriver::log_thread_created;
+
+// logging loop, needs to be static as it loops over backends created
+// by all all backends
+void AP_HAL::UARTDriver::logging_loop(void)
+{
+    while (true) {
+        hal.scheduler->delay(10);
+        for (auto *loginfo_ptr = backend_loginfos; loginfo_ptr != nullptr; loginfo_ptr = loginfo_ptr->next) {
+            auto &loginfo = *loginfo_ptr;
```

Comment:
```suggestion
        hal.scheduler->delay(10);
        AP_HAL::UARTDriver::LogInfo *loginfo_ptr;
        {
            WITH_SEMAPHORE(loginfo_creation_sem);
            loginfo_ptr = backend_loginfos;
        }
        for (; loginfo_ptr != nullptr; loginfo_ptr = loginfo_ptr->next) {
            auto &loginfo = *loginfo_ptr;
```

This is the correct way to do it for non-embedded platforms due to the relation with what the struct points to.

Reply:
What exactly are you trying to protect against here?

Reply:
There's no guarantee that another thread will see the contents of the loginfo struct before it sees the pointer to it. It's rather unlikely here but I think ArduPilot is a bit sleazy with locking and atomics in general...

Reply:
> There's no guarantee that another thread will see the contents of the loginfo struct before it sees the pointer to it. It's rather unlikely here but I think ArduPilot is a bit sleazy with locking and atomics in general and I would like to see it done better...

I can't see how the pointer is exposed to the other threads before the guts have been initialised....

And just generally, any time I need more nightmares on race conditions I go back to and read Paul McKenny's paper.  So I'm covered there.